lib/scholar/interpolation/cubic_spline.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Scholar.Interpolation.CubicSpline do
+  @moduledoc """
+  Cubic spline interpolation
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:coefficients, :x]}
+  defstruct [:coefficients, :x]
+
+  opts = [
+    extrapolate: [
+      required: false,
+      default: true,
+      type: :boolean,
+      doc: "if false, out-of-bounds x return NaN."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a cubic spline interpolation of the given `(x, y)` points
+  """
+  defn train(x, y) do
+    # https://en.wikiversity.org/wiki/Cubic_Spline_Interpolation
+    # Reference implementation in Scipy
+
+    x = Nx.as_type(x, :f64)
+    y = Nx.as_type(y, :f64)
```

Comment:
What do we force such type conversions? If the user wants f64 precision, then they must pass f64 in?